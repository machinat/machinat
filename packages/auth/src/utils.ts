import type { ServerResponse, IncomingMessage } from 'http';
import { posix as posixPath } from 'path';
import getRawBody from 'raw-body';
import {
  parse as parseCookie,
  serialize as serializeCookie,
  CookieSerializeOptions,
} from 'cookie';
import type {
  WithHeaders,
  AuthApiResponseBody,
  AuthApiErrorBody,
} from './types.js';

export const respondRedirect = (res: ServerResponse, url: string): void => {
  res.writeHead(302, { Location: url });
  res.end();
};

export const getCookies = (req: WithHeaders): null | Record<string, string> => {
  if (!req.headers.cookie) {
    return null;
  }
  return parseCookie(req.headers.cookie);
};

const SET_COOKIE = 'Set-Cookie';

export const setCookie = (
  res: ServerResponse,
  key: string,
  value: string,
  options: CookieSerializeOptions
): void => {
  const cookieDesc = serializeCookie(key, value, options);
  const cookiesAlreadySet = res.getHeader(SET_COOKIE);

  res.setHeader(
    SET_COOKIE,
    Array.isArray(cookiesAlreadySet)
      ? [...cookiesAlreadySet, cookieDesc]
      : typeof cookiesAlreadySet === 'string'
      ? [cookiesAlreadySet, cookieDesc]
      : cookieDesc
  );
};

export const isSubdomain = (domain: string, subdomain: string): boolean =>
  subdomain.endsWith(domain) &&
  (subdomain.length === domain.length ||
    subdomain[subdomain.length - domain.length - 1] === '.');

export const isSubpath = (
  path: null | string,
  subpath: null | string
): boolean =>
  !!(path && subpath) && !posixPath.relative(path, subpath).startsWith('..');

const CONTENT_TYPE_JSON = { 'Content-Type': 'application/json' };

export const respondApiOk = (
  res: ServerResponse,
  platform: string,
  token: string
): void => {
  res.writeHead(200, CONTENT_TYPE_JSON);
  const body: AuthApiResponseBody = { platform, token };
  res.end(JSON.stringify(body));
};

export const respondApiError = (
  res: ServerResponse,
  platform: undefined | string,
  code: number,
  reason: string
): void => {
  res.writeHead(code, CONTENT_TYPE_JSON);
  const body: AuthApiErrorBody = { platform, error: { code, reason } };
  res.end(JSON.stringify(body));
};

export const parseJsonBody = async (req: IncomingMessage): Promise<any> => {
  try {
    const rawBody = await getRawBody(req, { encoding: true });
    const body = JSON.parse(rawBody);

    return typeof body === 'object' ? body : null;
  } catch (err) {
    return null;
  }
};
