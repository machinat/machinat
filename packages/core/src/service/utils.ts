import invariant from 'invariant';
import {
  SOCIABLY_SERVICE_CONTAINER,
  SOCIABLY_SERVICE_PROVIDER,
  SOCIABLY_SERVICE_INTERFACE,
} from '../symbol.js';
import ServiceScope from './scope.js';
import ServiceMaker from './maker.js';
import ProvisionMap from './provisionMap.js';
import type {
  Interfaceable,
  ServiceRequirement,
  ServiceContainer,
  ServiceProvider,
  MaybeContainer,
} from './types.js';

export const isServiceContainer = <T>(
  target: MaybeContainer<T>,
): target is ServiceContainer<T, unknown[]> =>
  typeof target === 'function' &&
  '$$typeof' in target &&
  target.$$typeof === SOCIABLY_SERVICE_CONTAINER;

export const isServiceProvider = (
  target: any,
): target is ServiceProvider<unknown, unknown[]> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  target.$$typeof === SOCIABLY_SERVICE_PROVIDER;

export const maybeInjectContainer = <T>(
  scope: ServiceScope,
  maybeContainer: MaybeContainer<T>,
): T =>
  isServiceContainer(maybeContainer)
    ? scope.injectContainer(maybeContainer)
    : maybeContainer;

export const isInterfaceable = (target: any): target is Interfaceable<any> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  (target.$$typeof === SOCIABLY_SERVICE_INTERFACE ||
    target.$$typeof === SOCIABLY_SERVICE_PROVIDER);

export const createEmptyScope = (): ServiceScope =>
  new ServiceScope(new ServiceMaker(new ProvisionMap()), new Map());

export const polishServiceRequirement = <T>(
  dep: Interfaceable<T> | ServiceRequirement<Interfaceable<T>>,
): ServiceRequirement<Interfaceable<T>> => {
  if (isInterfaceable(dep)) {
    return { require: dep, optional: false };
  }

  invariant(
    isInterfaceable(dep?.require),
    `${(dep as any).name || String(dep)} is not a valid interface`,
  );

  return dep;
};

export const validateLifetime = (lifetime: string): void => {
  invariant(
    lifetime === 'singleton' ||
      lifetime === 'scoped' ||
      lifetime === 'transient',
    `${lifetime} is not valid service lifetime`,
  );
};
