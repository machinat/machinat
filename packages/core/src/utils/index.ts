export { default as compose } from './compose';
export { default as formatNode } from './formatNode';
export { default as getTimeId } from './getTimeId';
export { default as mixin } from './mixin';
export * from './isX';
