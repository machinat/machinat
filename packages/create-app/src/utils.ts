export const when =
  (shouldOutput: boolean) =>
  (strs: TemplateStringsArray, ...inputs: unknown[]): string => {
    if (!shouldOutput) {
      return '';
    }

    let outputStr = '';

    for (let i = 0; i < strs.length; i += 1) {
      outputStr += strs[i] + (inputs[i] || '');
    }

    return outputStr;
  };

export const polishFileContent = (
  input: null | string | boolean
): null | string | boolean => {
  if (typeof input !== 'string') {
    return input;
  }
  if (input === '') {
    return null;
  }

  let content = input;

  if (content[0] === '\n') {
    content = content.slice(1);
  }
  if (content[content.length - 1] !== '\n') {
    content = `${content}\n`;
  }

  return content;
};

export const getEmptyEnvs = (dotenvContent: string): Map<string, string[]> => {
  const lines = dotenvContent.split('\n');

  const result = new Map();
  let curPlatform = 'default';

  for (const line of lines) {
    if (line.match(/^[A-Z_]+=$/)) {
      const requiredEnv = line.slice(0, -1);

      const envs = result.get(curPlatform);
      if (envs) {
        envs.push(requiredEnv);
      } else {
        result.set(curPlatform, [requiredEnv]);
      }
    } else if (line.match(/^# .+$/)) {
      curPlatform = line.slice(2);
    }
  }

  return result;
};
