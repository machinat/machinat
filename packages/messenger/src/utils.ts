import { makeNativeComponent } from '@machinat/core/renderer';
import type {
  MessengerSegmentValue,
  MessageValue,
  SenderActionValue,
} from './types';
import { MESSENGER, API_PATH } from './constant';

const { hasOwnProperty } = Object.prototype;
export const isMessageEntry = (
  value: string | MessengerSegmentValue
): value is string | MessageValue | SenderActionValue =>
  typeof value === 'string' ||
  (typeof value === 'object' &&
    value !== null &&
    !hasOwnProperty.call(value, API_PATH));

export const makeMessengerComponent = makeNativeComponent(MESSENGER);
