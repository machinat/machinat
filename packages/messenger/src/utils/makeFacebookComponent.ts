import { makeNativeComponent } from '@sociably/core/renderer';
import { MESSENGER } from '../constant';

export default makeNativeComponent(MESSENGER);
