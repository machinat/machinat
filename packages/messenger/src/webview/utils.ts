/* eslint-disable import/prefer-default-export */
import type { ContextSupplement } from '@machinat/auth';
import MessengerChat from '../Chat';
import MessengerUser from '../User';
import type { MessengerAuthContext, MessengerAuthData } from './types';

export const supplementContext = (
  data: MessengerAuthData
): ContextSupplement<MessengerAuthContext> => {
  const {
    user: userId,
    page: pageId,
    chat: { type: chatType, id: chatId },
    client,
  } = data;
  return {
    user: new MessengerUser(pageId, userId),
    channel: new MessengerChat(pageId, { id: chatId }, chatType),
    pageId,
    clientType: client,
  };
};
