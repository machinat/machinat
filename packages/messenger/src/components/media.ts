import { makeUnitSegment } from '@machinat/core/renderer';
import type { UnitSegment } from '@machinat/core/renderer';
import { makeMessengerComponent } from '../utils';
import {
  ATTACHMENT_ASSET_TAG,
  ATTACHMENT_DATA,
  ATTACHMENT_INFO,
} from '../constant';
import type { MessageValue, MessengerComponent } from '../types';

/**
 * @category Props
 */
export type MediaProps = {
  /**
   * URL of the file to upload. Max file size is 25MB (after encoding). A
   * Timeout is set to 75 sec for videos and 10 secs for every other file type.
   */
  url: string;
  /**
   * Set to true to make the saved asset sendable to other message recipients.
   * Defaults to false.
   */
  isReusable?: boolean;
  /**
   * Set to true to make the saved asset sendable to other message recipients.
   * Defaults to false.
   */
  attachmentId?: string;

  attachmentAssetTag?: string;
  attachmentFileData?: string | Buffer | ReadableStream;
  attachmentFileInfo?: {
    filename?: string;
    filepath?: string;
    contentType?: string;
    knownLength?: number;
  };
};

const mediaFactory = (
  name: string,
  type: string
): MessengerComponent<MediaProps, UnitSegment<MessageValue>> => {
  const container = {
    [name]: (node, path) => {
      const {
        url,
        isReusable,
        attachmentId,
        attachmentAssetTag,
        attachmentFileData,
        attachmentFileInfo,
      } = node.props;

      return [
        makeUnitSegment(node, path, {
          message: {
            attachment: {
              type,
              payload: {
                url,
                is_reusable: isReusable,
                attachment_id: attachmentId,
              },
            },
          },
          [ATTACHMENT_ASSET_TAG]: attachmentAssetTag,
          [ATTACHMENT_DATA]: attachmentFileData,
          [ATTACHMENT_INFO]: attachmentFileInfo,
        }),
      ];
    },
  };

  return makeMessengerComponent(container[name]);
};

/**
 * The log out button triggers the account unlinking flow.
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official send API [doc](https://developers.facebook.com/docs/messenger-platform/send-messages)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/send-api).
 */
export const Image = mediaFactory('Image', 'image');

/**
 * The log out button triggers the account unlinking flow.
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official send API [doc](https://developers.facebook.com/docs/messenger-platform/send-messages)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/send-api).
 */
export const Video = mediaFactory('Video', 'video');

/**
 * The log out button triggers the account unlinking flow.
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official send API [doc](https://developers.facebook.com/docs/messenger-platform/send-messages)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/send-api).
 */
export const Audio = mediaFactory('Audio', 'audio');

/**
 * The log out button triggers the account unlinking flow.
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official send API [doc](https://developers.facebook.com/docs/messenger-platform/send-messages)
 *   and [reference](https://developers.facebook.com/docs/messenger-platform/reference/send-api).
 */
export const File = mediaFactory('File', 'file');
