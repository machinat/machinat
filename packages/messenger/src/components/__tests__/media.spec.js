import Machinat from '@machinat/core';
import { isNativeType } from '@machinat/core/utils';
import Renderer from '@machinat/core/renderer';

import { Image, Video, Audio, File } from '../media';

const renderer = new Renderer('messenger', () => null);

describe('media Components', () => {
  test.each([Image, Video, Audio, File])('is valid root Component', (Media) => {
    expect(typeof Media).toBe('function');
    expect(isNativeType(<Media />)).toBe(true);
    expect(Media.$$platform).toBe('messenger');
  });

  it('match snapshot', async () => {
    const segments = await renderer.render(
      <>
        <Image url="http://this.is/a/picture" reusable />
        <Image attachmentId="_I_am_Image_" />
        <Video url="http://this.is/a/video" reusable />
        <Video attachmentId="_I_am_Video_" />
        <Audio url="http://this.is/an/audio" reusable />
        <Audio attachmentId="_I_am_Audio_" />
        <File url="http://this.is/a/file" reusable />
        <File attachmentId="_I_am_File_" />
      </>
    );
    expect(segments).toMatchSnapshot();
    expect(segments.map((seg) => seg.value)).toMatchInlineSnapshot(`
      Array [
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": true,
                "url": "http://this.is/a/picture",
              },
              "type": "image",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": "_I_am_Image_",
                "is_reusable": undefined,
                "url": undefined,
              },
              "type": "image",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": true,
                "url": "http://this.is/a/video",
              },
              "type": "video",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": "_I_am_Video_",
                "is_reusable": undefined,
                "url": undefined,
              },
              "type": "video",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": true,
                "url": "http://this.is/an/audio",
              },
              "type": "audio",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": "_I_am_Audio_",
                "is_reusable": undefined,
                "url": undefined,
              },
              "type": "audio",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": undefined,
                "is_reusable": true,
                "url": "http://this.is/a/file",
              },
              "type": "file",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
        Object {
          "message": Object {
            "attachment": Object {
              "payload": Object {
                "attachment_id": "_I_am_File_",
                "is_reusable": undefined,
                "url": undefined,
              },
              "type": "file",
            },
          },
          Symbol(attatchment_asset_tag.messenger.machinat): undefined,
          Symbol(attachment_data.messenger.machinat): undefined,
          Symbol(attachment_info.messenger.machinat): undefined,
        },
      ]
    `);
  });
});
