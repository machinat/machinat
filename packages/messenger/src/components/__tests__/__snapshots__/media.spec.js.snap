// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`media Components match snapshot 1`] = `
Array [
  Object {
    "node": <Image
      reusable={true}
      url="http://this.is/a/picture"
    />,
    "path": "$::0",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": undefined,
            "is_reusable": true,
            "url": "http://this.is/a/picture",
          },
          "type": "image",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
  Object {
    "node": <Image
      attachmentId="_I_am_Image_"
    />,
    "path": "$::1",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": "_I_am_Image_",
            "is_reusable": undefined,
            "url": undefined,
          },
          "type": "image",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
  Object {
    "node": <Video
      reusable={true}
      url="http://this.is/a/video"
    />,
    "path": "$::2",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": undefined,
            "is_reusable": true,
            "url": "http://this.is/a/video",
          },
          "type": "video",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
  Object {
    "node": <Video
      attachmentId="_I_am_Video_"
    />,
    "path": "$::3",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": "_I_am_Video_",
            "is_reusable": undefined,
            "url": undefined,
          },
          "type": "video",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
  Object {
    "node": <Audio
      reusable={true}
      url="http://this.is/an/audio"
    />,
    "path": "$::4",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": undefined,
            "is_reusable": true,
            "url": "http://this.is/an/audio",
          },
          "type": "audio",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
  Object {
    "node": <Audio
      attachmentId="_I_am_Audio_"
    />,
    "path": "$::5",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": "_I_am_Audio_",
            "is_reusable": undefined,
            "url": undefined,
          },
          "type": "audio",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
  Object {
    "node": <File
      reusable={true}
      url="http://this.is/a/file"
    />,
    "path": "$::6",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": undefined,
            "is_reusable": true,
            "url": "http://this.is/a/file",
          },
          "type": "file",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
  Object {
    "node": <File
      attachmentId="_I_am_File_"
    />,
    "path": "$::7",
    "type": "unit",
    "value": Object {
      "message": Object {
        "attachment": Object {
          "payload": Object {
            "attachment_id": "_I_am_File_",
            "is_reusable": undefined,
            "url": undefined,
          },
          "type": "file",
        },
      },
      Symbol(attatchment_asset_tag.messenger.sociably): undefined,
      Symbol(attachment_data.messenger.sociably): undefined,
      Symbol(attachment_info.messenger.sociably): undefined,
    },
  },
]
`;
